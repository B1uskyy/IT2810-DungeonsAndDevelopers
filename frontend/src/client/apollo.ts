import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});

const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
