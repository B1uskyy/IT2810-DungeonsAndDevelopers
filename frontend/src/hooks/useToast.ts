import { useContext } from 'react';
import { ToastContext } from '../context/ToastContext.tsx';

export const useToast = () => {
  const context = useContext(ToastContext);
  if (!context) {
    // Return a fallback or throw an error
    return {
      showToast: () => {
        console.warn('Toast not available in this context.');
      },
      toastId: { current: null },
    };
  }
  return context;
};
