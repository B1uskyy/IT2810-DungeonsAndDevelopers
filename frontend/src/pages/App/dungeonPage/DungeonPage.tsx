import ModeEditOutlineIcon from '@mui/icons-material/ModeEditOutline';
import SaveIcon from '@mui/icons-material/Save';
import Monsters from '../../../data/mockup.ts';
import '../../../components/MonsterCard/MonsterCard.css';
import Navbar from '../../../components/Navbar/Navbar.tsx';
import React, { useEffect, useRef, useState } from 'react';

export default function DungeonPage() {
  const [monsters, setMonsters] = useState(Monsters.results.slice(0, 6));
  const dungeonNamePlaceholder = 'Enter dungeon name';
  const [dungeonName, setDungeonName] = useState(() => {
    const savedName = localStorage.getItem('dungeonName');
    return savedName || dungeonNamePlaceholder;
  });
  const [isEditing, setIsEditing] = useState(false);
  // const [isOpen, setIsOpen] = useState(false);

  useEffect(() => {
    if (dungeonName != dungeonNamePlaceholder) {
      localStorage.setItem('dungeonName', dungeonName);
    }
  }, [dungeonName]);

  const dungeonNameRef = useRef<HTMLInputElement>(null);

  const handleDungeonClick = () => {
    setIsEditing(true);
  };

  const handleSaveClick = () => {
    const currentValue = dungeonNameRef.current?.value.trim() || '';

    if (currentValue === '' || currentValue === dungeonNamePlaceholder) {
      alert('Dungeon name cannot be empty!');
      setDungeonName(dungeonNamePlaceholder);
    } else {
      setDungeonName(dungeonName);
    }
    setIsEditing(false);
  };

  const handleBlur = () => {
    handleSaveClick();
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    let newName = e.target.value;

    if (newName.length > 20) {
      newName = newName.substring(0, 20);
    }
    setDungeonName(newName);
  };

  // For modal (each monster):
  // const handleClose = () => {
  //   setIsOpen(!isOpen);
  // };

  return (
    <section className="min-h-screen flex flex-col bg-storm_giant bg-cover bg-center bg-no-repeat relative">
      <Navbar />
      <main className="flex-grow flex flex-col items-center justify-center relative z-10">
        <article className="w-full max-w-7xl mx-auto px-4 py-8">
          <header className="mb-12 text-center">
            <h1 className="text-5xl md:text-6xl text-white font-bold mb-4">Dungeon creator</h1>
          </header>
          <section className="bg-customGray bg-opacity-50 p-8 rounded-lg shadow-lg">
            <header className="flex flex-row items-center justify-center gap-4 mb-8">
              {isEditing ? (
                <input
                  ref={dungeonNameRef}
                  type="text"
                  defaultValue={dungeonName === dungeonNamePlaceholder ? '' : dungeonName}
                  onChange={handleChange}
                  onBlur={handleBlur}
                  autoFocus
                  className="focus:outline-customRed focus:ring-customRed  text-xl font-bold text-center"
                />
              ) : (
                <h2 className="text-3xl md:text-4xl text-white font-semibold">{dungeonName}</h2>
              )}

              <button
                className="p-2 rounded-lg hover:bg-white transition-colors duration-300 group"
                aria-label="Edit dungeon name"
                onClick={isEditing ? handleSaveClick : handleDungeonClick}
              >
                {isEditing ? (
                  <SaveIcon fontSize="large" className="text-white group-hover:text-black duration-100" />
                ) : (
                  <ModeEditOutlineIcon fontSize="large" className="text-white group-hover:text-black duration-100" />
                )}
              </button>
            </header>
            {/*
            Add difficulty section -> check figma
            */}
            <ul className="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-6 list-none">
              {monsters.map((monster, index) => (
                <li
                  key={index}
                  className="bg-customRed p-4 rounded-lg shadow-md hover:shadow-lg transition-shadow duration-300"
                >
                  <h3 className="text-white text-lg">{monster.name}</h3>
                </li>
              ))}
            </ul>
          </section>
        </article>
      </main>
    </section>
  );
}
