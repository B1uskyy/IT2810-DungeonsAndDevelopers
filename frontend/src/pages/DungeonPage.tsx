import ModeEditOutlineIcon from '@mui/icons-material/ModeEditOutline';

import Monsters from '../data/mockup.ts';
import { useEffect, useState } from 'react';
import '../components/MonsterCard/MonsterCard.css';

export default function DungeonPage() {
  const name: string = 'The sunken citadel';
  const [monsters, setMonsters] = useState(Monsters.results);

  useEffect(() => {
    setMonsters(monsters.slice(0, 6));
  }, [monsters]);
  return (
    <>
      <main className="flex flex-col text-center gap-20">
        <header className="mt-8 flex flex-col gap-6 items-center">
          {/*
          Add image wiht object-cover:fit
          */}
          <h1 className="text-6xl">Dungeon creator</h1>
          <span className="flex flex-row gap-4">
            <h4 className="text-4xl">{name}</h4>
            <button className="btn btn-reset hover:cursor-pointer rounded-lg hover:bg-customRed hover:text-white">
              <ModeEditOutlineIcon fontSize="large" />
            </button>
          </span>
        </header>
        <section className="flex flex-col items-center">
          {/*<h5 className="text-2xl w-1/2">Your dungeon monsters</h5>*/}
          {/*
          Add difficulty section -> check figma
          */}
        </section>
        <section className="grid grid-cols-3 items-center justify-center">
          {monsters.map((monster, index) => (
            <div key={index} className="card">
              <p>{monster.name}</p>
            </div>
          ))}
        </section>
      </main>
    </>
  );
}
