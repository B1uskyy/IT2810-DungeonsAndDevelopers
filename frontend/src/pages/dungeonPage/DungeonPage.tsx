import '../../components/MonsterCard/MonsterCard.css';
import Navbar from '../../components/Navbar/Navbar.tsx';
import { useState } from 'react';
import DungeonHeader from '../../components/Dungeon/DungeonHeader.tsx';
import CustomInput from '../../components/Dungeon/CustomInput.tsx';
import DungeonMonsterGrid from '../../components/Dungeon/DungeonMonsterGrid.tsx';

export default function DungeonPage() {
  const [dungeonName, setDungeonName] = useState(() => {
    const savedName = localStorage.getItem('dungeonName');
    return savedName || 'Enter dungeon name';
  });

  const handleSaveDungeonName = (newName: string) => {
    setDungeonName(newName);
    localStorage.setItem('dungeonName', newName);
  };

  return (
    <section className="min-h-screen flex flex-col bg-storm_giant bg-cover bg-center bg-no-repeat relative">
      <Navbar />
      <main className="flex-grow flex flex-col items-center justify-center relative z-10 ">
        <article className="flex flex-col gap-16 items-center w-full max-w-7xl mx-auto">
          <DungeonHeader />
          <section className="bg-customGray bg-opacity-80 p-8 rounded-lg shadow-lg w-2/3 tablet:w-10/12">
            <CustomInput
              placeholder="Enter dungeon name"
              inputName="Dungeon Name"
              value={dungeonName} 
              onSave={handleSaveDungeonName}
            />
            <DungeonMonsterGrid /> 
          </section>
        </article>
      </main>
    </section>
  );
}
