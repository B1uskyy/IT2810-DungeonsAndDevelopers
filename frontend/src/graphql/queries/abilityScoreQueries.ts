import { gql } from '@apollo/client';

export const GET_ABILITYSCORES = gql`
  query GetAbilityScores($offset: Int, $limit: Int) {
    abilities(offset: $offset, limit: $limit) {
      abilities {
        index
        name
        skills
      }
      totalAbilities
    }
  }
`;
