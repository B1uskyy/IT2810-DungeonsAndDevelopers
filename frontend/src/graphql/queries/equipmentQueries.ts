import { gql } from '@apollo/client';

export const GET_EQUIPMENT_SUGGESTIONS = gql`
  query GetEquipmentSuggestions($searchTerm: String!) {
    equipments(searchTerm: $searchTerm, suggestionsOnly: true) {
      equipments {
        id
        name
      }
      totalCount
    }
  }
`;

export const GET_EQUIPMENTS = gql`
  query GetEquipments($searchTerm: String, $offset: Int, $limit: Int) {
    equipments(searchTerm: $searchTerm, offset: $offset, limit: $limit) {
      equipments {
        id
        index
        name
        category
        value
      }
      totalCount
    }
  }
`;
