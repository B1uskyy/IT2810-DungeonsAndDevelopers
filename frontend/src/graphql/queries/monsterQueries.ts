import { gql } from '@apollo/client';

export const GET_MONSTERS = gql`
  query GetMonsters(
    $searchTerm: String
    $offset: Int
    $limit: Int
    $types: [String!]
    $minHp: Int
    $maxHp: Int
    $sortOption: String
  ) {
    monsters(
      searchTerm: $searchTerm
      offset: $offset
      limit: $limit
      types: $types
      minHp: $minHp
      maxHp: $maxHp
      sortOption: $sortOption
    ) {
      monsters {
        id
        name
        size
        type
        alignment
        hit_points
        image
        averageDifficulty
        reviewsCount
      }
      totalMonsters
      minHp
      maxHp
    }
  }
`;

export const GET_MONSTER_TYPE_COUNTS = gql`
  query GetMonsterTypeCounts($minHp: Int, $maxHp: Int) {
    monsterTypeCounts(minHp: $minHp, maxHp: $maxHp) {
      type
      count
    }
  }
`;

export const GET_MONSTER_HP_RANGE = gql`
  query GetMonsterHpRange {
    monsterHpRange {
      minHp
      maxHp
    }
  }
`;

export const GET_MONSTER_SUGGESTIONS = gql`
  query GetMonsterSuggestions($searchTerm: String, $types: [String!], $minHp: Int, $maxHp: Int) {
    monsters(searchTerm: $searchTerm, types: $types, minHp: $minHp, maxHp: $maxHp) {
      monsters {
        id
        name
      }
    }
  }
`;

export const GET_MONSTER_REVIEWS = gql`
  query GetMonsterReviews($monsterId: ID!) {
    monster(id: $monsterId) {
      id
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;
