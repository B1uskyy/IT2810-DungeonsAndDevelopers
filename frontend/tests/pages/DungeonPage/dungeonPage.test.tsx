import { render, screen, waitFor } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { MockedProvider } from '@apollo/client/testing';
import DungeonPage from '../../../src/pages/mainPages/dungeonPage.tsx';
import { GET_USER_DUNGEON, UPDATE_DUNGEON_NAME } from '../../../../backend/src/graphql/queries.ts';
import { useToast } from '../../../src/hooks/useToast.ts';
import { AuthContext } from '../../../src/context/AuthContext.tsx';
import { BrowserRouter } from 'react-router-dom';
import { expect } from 'vitest';

const mockShowToast = vi.fn();
vi.mock('../hooks/useToast.ts', () => ({
  useToast: () => ({
    showToast: mockShowToast,
  }),
}));

const mockMonsters = [
  { id: '1', name: 'Goblin', hp: 30, type: 'Beast', size: 'Small', img: '' },
  { id: '2', name: 'Orc', hp: 50, type: 'Humanoid', size: 'Medium', img: '' },
];

const mocks = [
  {
    request: {
      query: GET_USER_DUNGEON,
      variables: { userId: '1' },
    },
    result: {
      data: {
        user: {
          dungeonName: 'Example dungeon',
          favoritedMonsters: mockMonsters,
        },
      },
    },
  },
  {
    request: {
      query: UPDATE_DUNGEON_NAME,
      variables: { userId: '1', dungeonName: 'New dungeon' },
    },
    result: {
      data: {
        updateDungeonName: {
          dungeonName: 'New dungeon',
        },
      },
    },
  },
  {
    request: {
      query: GET_USER_DUNGEON,
      variables: { userId: '1' },
    },
    result: {
      data: {
        user: {
          dungeonName: 'New dungeon',
          favoritedMonsters: mockMonsters,
        },
      },
    },
  },
];

describe('DungeonPage', () => {
  const userId = '1';
  const token = 'mock-token'; // Mock token
  const userName = 'Mock User'; // Mock user name

  beforeEach(() => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AuthContext.Provider value={{ userId, token, userName, login: vi.fn(), logout: vi.fn() }}>
          <BrowserRouter>
            <DungeonPage />
          </BrowserRouter>
        </AuthContext.Provider>
      </MockedProvider>
    );
  });

  it('renders correctly and matches snapshot', async () => {
    // Wait for the data to be loaded and elements to appear
    await waitFor(() => expect(screen.getByTestId('dungeon-name')).toBeInTheDocument());
    expect(screen.getByText('My Dungeon')).toBeInTheDocument();
    expect(screen.getByText('No monsters in dungeon')).toBeInTheDocument();

    // Snapshot testing
    expect(screen.getByRole('main')).toMatchSnapshot();
  });

  it('updates dungeon name on input save', async () => {
    const user = userEvent.setup();
    const editButton = screen.getByRole('button', { name: 'Edit Dungeon Name' });

    await user.click(editButton);
    // Use userEvent to change the input and simulate pressing Enter

    const input = screen.getByRole('textbox', { name: 'Edit name' });
    await user.clear(input);
    await user.type(input, 'New dungeon');

    const saveButton = screen.getByRole('button', { name: 'Save Dungeon Name' });

    await user.click(saveButton);

    // Check if the toast notification is shown
    await waitFor(() =>
      expect(mockShowToast).toHaveBeenCalledWith({
        message: 'Dungeon name updated successfully',
        type: 'success',
        duration: 3000,
      })
    );
    await waitFor(() => {
      expect(screen.getByTestId('dungeon-name')).toHaveTextContent('New dungeon');
    });
  });

  it('displays total HP when monsters are present', async () => {
    // Update the mock data to include monsters
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AuthContext.Provider value={{ userId, userName, login: vi.fn(), logout: vi.fn(), token }}>
          <BrowserRouter>
            <DungeonPage />
          </BrowserRouter>
        </AuthContext.Provider>
      </MockedProvider>
    );

    // Wait for the stats to update
    await waitFor(() => {
      expect(screen.getByText(/Total Dungeon HP: 80/i)).toBeInTheDocument();
    });
  });

  it('handles adding a monster to the dungeon', async () => {
    // Render the monster card
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <AuthContext.Provider value={{ userId, userName, login: vi.fn(), logout: vi.fn(), token }}>
          <BrowserRouter>
            <DungeonPage />
          </BrowserRouter>
        </AuthContext.Provider>
      </MockedProvider>
    );

    // Simulate clicking the add button on the monster card using userEvent
    const addButton = screen.getByLabelText('Click to add to dungeon');
    await userEvent.click(addButton);

    // Verify if the toast is shown for adding monster
    await waitFor(() =>
      expect(mockShowToast).toHaveBeenCalledWith({
        message: 'Goblin added to dungeon',
        type: 'success',
        duration: 3000,
      })
    );
  });
});
