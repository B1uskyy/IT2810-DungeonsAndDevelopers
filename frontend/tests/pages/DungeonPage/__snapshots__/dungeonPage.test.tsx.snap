// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DungeonPage > renders correctly and matches snapshot 1`] = `
<main
  class="main before:bg-dungeon xl:h-screen xl:overflow-hidden"
>
  <div
    class="black-overlay"
  />
  <div
    class="wrapper min-w-[70%] mt-[10vh] h-full !justify-start"
  >
    <section
      class="flex flex-row items-center justify-center gap-4 "
    >
      <h2
        aria-label="Name of your dungeon"
        class=" text-6xl md:text-5xl lg:text-4xl xl:text-3xl 2xl:text-2xl"
        data-testid="dungeon-name"
      >
        My Dungeon
      </h2>
      <button
        aria-label="Edit Dungeon Name"
        class=" p-[2px] lg:p-[5px] rounded-lg bg-customRed hover:bg-transparent border-2 border-customRed hover:border-customRed transition-colors duration-100 group"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium text-white group-hover:text-customRed duration-100 lg:!w-6 lg:!h-6 xl:!w-8 xl:!h-8 css-20bmp1-MuiSvgIcon-root"
          data-testid="ModeEditOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 5.63l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83c.39-.39.39-1.02 0-1.41"
          />
        </svg>
      </button>
    </section>
    <div
      class="grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3 gap-x-10 xl:gap-x-28 gap-y-5 place-items-center transition-opacity duration-500 p-5"
    >
      <div
        class="flex items-center justify-center h-full w-full col-span-full text-center"
      >
        <p
          class="sub-header"
        >
          No monsters in dungeon
        </p>
      </div>
    </div>
  </div>
</main>
`;
